sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -1067,8 +1063,11 @@ private void checkForFailures(String message) throws IOException {
 
       @Teardown
       public void teardown() throws Exception {
-        if (producer != null && producer.getOutstandingRecordsCount() > 0) {
-          producer.flushSync();
+        if (producer != null) {
+          if (producer.getOutstandingRecordsCount() > 0) {
+            producer.flushSync();
+          }
+          producer.destroy();
         }
         producer = null;
```

Comment:
@aromanenko-dev After stumbling across this intended change, and as you reviewed the [originating MR](https://github.com/apache/beam/pull/9640) where the code is originating I am wondering, how this is working at all.

Iiuc this var 'producer' is static on KinesisWriterFn. Couldn't setting this to 'null' in @Teardown of some DoFn instance result in random NPEs thrown in some other DoFn currently running @ProcessElement or @FinishBundle? As it seems, it does only work because unsetting producer simply does not cross some memory fences?

So probably I have a misunderstanding of the inner workings here. What are you thoughts about this?

Reply:
👍 definitely wondering myself how this doesn't blow up randomly :/

Reply:
Finally getting back to this. I think the reason this worked is `producer.flushSync()`. With that a busy producer likely won't get destroyed. Really sketchy ...
```
    public void flushSync() {
        while (getOutstandingRecordsCount() > 0) {
            flush();
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) { }
        }
    }
    ```

Reply:
> 

One reason that could explain that it works is that some execution engines write at parallelism one resulting into a single instance of the write DoFn

Reply:
I was wondering about that @echauchot .... To they execution engine a writer is just a DoFn, right? How would it know to reduce parallelism?

Reply:
To the execution engine, a writer is a native `map` that calls the Beam `ParDo` that runs the `DoFn`. The engine instanciates the `map` and for each map there is only one `DoFn`, so if the engine choses parallelism 1 for `map`, it will result in a single `DoFn`. In a nutshell, Beam never controls parallelism, it is always the choice of the engine

Reply:
Thats absolutely clear 👍 

> some execution engines write at parallelism one resulting into a single instance of the write DoFn

Though my question is how could the execution engine distinguish a DoFn that implements a write from any other DoFn to reduce parallelism and accomplish the above?

Reply:
I checked the runner code. What happens is that the `Write` returns a `PDone` (empty `PCollection`). the Write is executed with a native `map` and the output of the map is the last native `Dataset` of the native (generated) pipeline on which we call a passthrough native `writer` but the actual write operation has already been done in reality (when we executed the map). Conclusion the engine will set the parallelism on the writer that does nothing. So I was wrong, the parallelism of the writing operation will be the one set by the engine for the final map of the DAG.

Reply:
Thanks a lot for checking, that really helps to better understand things @echauchot 🙏